### ISSUE TITLE
Moving code from one file to another breaks its crate imports

### ISSUE THREAD

**Issue by cardiffman** [rust-lang/rust#29737]
If you have successful usage of xml_rs in main.rs

    // main.rs
    extern crate xml;
    use xml::reader::{EventReader, XmlEvent};
    /* and here comes some XML-handling code */

And after working with it for a bit you move the XML code to xml_usecase.rs 

    // main.rs
    mod xml_usecase;
    use xml_usecase::*;

    // xml_usecase.rs
    extern crate xml;
    use xml::reader::{EventReader, XmlEvent};
    /* and here comes some XML-handling code */

The same exact text will now spew error messages. The files by the way are in the same src directory.

Given the steep slopes a new rust developer must climb to code the little things, it seems like this is an unnecessary barrier to getting code from the point where it's all stuffed in a 1000-line main.rs to where the code is divided into logical pieces. 

[This S.O. answer](http://stackoverflow.com/a/27653246) says what to do about it. But one of the strategies means that code files can't be moved around without touching it inside the files. And the other one, putting the crate imports in the root, means that besides putting things in the cargo.toml and at the site of use, we are putting things in a file that should not know so much about so many things.

main.rs does need to "point to" xml_usecase.rs but it isn't very fluid to have to have xml_usecase.rs contain so much modification to its boiler-plate compared to how it looked when it cohabited with main.rs.

**Comment by nikomatsakis** [comment]
This is just how it is. Extern crate declarations really ought to be placed at the root. `use` statements are always absolute and we really can't change this -- both because of backwards compatibility, and because the name resolution algorithm itself becomes rather horribly complex otherwise. The reason is that when the root of your path depends on what's in scope, strange relationships arise between `pub use` and so forth.

**Comment by cardiffman** [comment]
Laying aside type-safety, memory safety and so on and speaking strictly of the build system and modules, what am I winning in exchange? In C++ I can do `g++ -o app main.cpp xml_usecase.cpp -lexpat` and "up-arrow" is my makefile. The `-lexpat` is a wart on C++, right? When I start to add other things to my app like HTTP further `-l` options come along and you have to use `pkg-config` even if you're using a makefile. In a clean-sheet design that sort of information could be embedded in xml_usecase.rs instead of having to be remembered in command history or a build script. In rust it seems like I have to work harder at an earlier stage, and the idea that  xml_usecase.rs provides an abstraction is not attained as compensation if I have to maintain a list of crate names in my root `.rs` file.

**Comment by cardiffman** [comment]
When I first heard about rust I had heard MLTon and read the MLTon papers, and I was excited that a language of that type was going to be tried so seriously. And now it seems like it has an ecosystem and everything.

**Comment by nrc** [comment]
This is ought to be "easily" solvable with a refactoring tool. If anyone wants to try to write it, I'd be happy to mentor/help out.

**Comment by oli-obk** [comment]
how about linting against `extern crate xyz;` outside the root module? and also improving the unresolved import lint to check if it would work if it had `self::` or `super::` prepended and suggest that.

The latter is not different from suggesting static functions vs methods vs function-type fields on structs.

**Comment by cardiffman** [comment]
I like the suggestion idea but (my-other-machine problem) I would say I vaguely remember a suggestion along those lines was emitted. It may have been, "Hey try super or self."

If I'm prototyping a feature by just dumping it in the first convenient file, I don't want what I'm going to suggest to be on all the time but, if there was a way to lint a type or set of types for the work that you'll have to do to move them out to a module that might be a convenience. Not only the crate references but (an issue I excluded from the scope of this report) the `pub`'s I had to add once I had moved the code. 

My above suggestion could help with a general rust compiler pattern: The compiler hardly ever tells me all that's wrong with my code if I am compiling it for the first time after an attempt to refactor it.
